File: apps/Ingestion Plane/Quarry-v2/pkg/quarrycontracts/output.go
package quarrycontracts

import "time"

type NormalizedOutput struct {
	RunID       ID             `json:"run_id"`
	URL         URLTriple      `json:"url"`
	Status      uint16         `json:"status"`
	FetchedAt   time.Time      `json:"fetched_at"`
	Fingerprint string         `json:"fingerprint"`
	Formats     OutputFormats  `json:"formats"`
	Change      ChangeInfo     `json:"change"`
	Metadata    PageMetadata   `json:"metadata"`
	Driver      DriverInfo     `json:"driver"`
}

type URLTriple struct {
	Requested string  `json:"requested"`
	Final     string  `json:"final"`
	Canonical *string `json:"canonical,omitempty"`
}

type OutputFormats struct {
	HTML       *FormatRef  `json:"html,omitempty"`
	Markdown   *FormatRef  `json:"markdown,omitempty"`
	Raw        *FormatRef  `json:"raw,omitempty"`
	Links      []Link      `json:"links,omitempty"`
	Screenshot *FormatRef  `json:"screenshot,omitempty"`
	PDF        *FormatRef  `json:"pdf,omitempty"`
	Extract    *ExtractRef `json:"extract,omitempty"`
}

type FormatRef struct {
	ArtifactID ID     `json:"artifact_id"`
	Bytes      uint64 `json:"bytes"`
}

type ExtractRef struct {
	ArtifactID ID      `json:"artifact_id"`
	SchemaID   *string `json:"schema_id,omitempty"`
}

type Link struct {
	Href string  `json:"href"`
	Text *string `json:"text,omitempty"`
	Rel  *string `json:"rel,omitempty"`
}

type ChangeInfo struct {
	Status           ChangeStatus `json:"status"`
	PrevFingerprint  *string      `json:"prev_fingerprint,omitempty"`
}

type ChangeStatus string
const (
	ChangeNew       ChangeStatus = "new"
	ChangeChanged   ChangeStatus = "changed"
	ChangeUnchanged ChangeStatus = "unchanged"
)

type PageMetadata struct {
	Title       *string `json:"title,omitempty"`
	Lang        *string `json:"lang,omitempty"`
	ContentType *string `json:"content_type,omitempty"`
}

type DriverInfo struct {
	Kind       DriverKind `json:"kind"`
	DurationMs uint64     `json:"duration_ms"`
}

type DriverKind string
const (
	DriverStatic  DriverKind = "static"
	DriverBrowser DriverKind = "browser"
	DriverTLS     DriverKind = "tls"
)

File: apps/Ingestion Plane/Quarry-v2/.claude/worktrees/vigilant-banach-6933aa/crates/quarry-core/src/output.rs
//! Normalized output envelope (runtime → edge/control). Mirrors CONTRACTS §9.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::ids::kinds;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NormalizedOutput {
    pub run_id: kinds::RunKind,
    pub url: UrlTriple,
    pub status: u16,
    pub fetched_at: DateTime<Utc>,
    pub fingerprint: String,
    pub formats: OutputFormats,
    pub change: ChangeInfo,
    pub metadata: PageMetadata,
    pub driver: DriverInfo,
    /// Cycle 21 / cluster #2 — runtime stamps the Determinism mode +
    /// policy fingerprint + identity hash so downstream consumers can
    /// audit which RunPolicy produced the artifact. `None` for legacy
    /// outputs produced before P21 wiring. Always populated by the
    /// current PageRunner.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub determinism: Option<DeterminismStamp>,
}

/// Compact policy/identity stamp embedded in [`NormalizedOutput`]. The
/// fields mirror [`crate::contracts`] / quarry-runtime's `policy.rs`
/// types but live in `quarry-core` so consumers don't need to depend
/// on the runtime crate to interpret the stamp.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeterminismStamp {
    /// `"strict" | "best_effort" | "off"`.
    pub mode: String,
    /// `policy_fingerprint(&RunPolicy)` — e.g. `"blake3:..."`.
    pub policy_fp: String,
    /// `record_determinism_inputs(url, &policy).id` — `"dq:..."`.
    pub identity_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UrlTriple {
    pub requested: String,
    pub final_url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub canonical: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct OutputFormats {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub html: Option<FormatRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub markdown: Option<FormatRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub raw: Option<FormatRef>,
    #[serde(default)]
    pub links: Vec<Link>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub screenshot: Option<FormatRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pdf: Option<FormatRef>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extract: Option<ExtractRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FormatRef {
    pub artifact_id: kinds::ArtifactKind,
    pub bytes: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractRef {
    pub artifact_id: kinds::ArtifactKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub schema_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Link {
    pub href: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rel: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChangeInfo {
    pub status: ChangeStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prev_fingerprint: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ChangeStatus {
    New,
    Changed,
    Unchanged,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PageMetadata {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lang: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub content_type: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DriverInfo {
    pub kind: DriverKind,
    pub duration_ms: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub profile: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub live_view_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recording_id: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DriverKind {
    Static,
    Browser,
    Tls,
}
